pkg/oci/oci.go

```diff
@@ -892,50 +887,45 @@ func ensureImage(ctx context.Context, imageStore oras.Target, image string, imgO
 		}
 	}
 
-	if len(imgOpts.AllowedRegistries) > 0 {
-		targetImage, err := normalizeImageName(image)
-		if err != nil {
-			return fmt.Errorf("normalizing image: %w", err)
-		}
-
-		image := targetImage.String()
+	if len(imgOpts.AllowedGadgets) > 0 {
 		found := false
 
-		for _, registry := range imgOpts.AllowedRegistries {
-			found = strings.HasPrefix(image, registry)
-			if found {
-				break
-			}
-		}
-
-		if !found {
-			return fmt.Errorf("%s not originating from allowed registries: %v", image, strings.Join(imgOpts.AllowedRegistries, ", "))
-		}
-	}
-
-	if len(imgOpts.AllowedDigests) > 0 {
-		targetImage, err := normalizeImageName(image)
+		normalizedImage, err := normalizeImageName(image)
 		if err != nil {
 			return fmt.Errorf("normalizing image: %w", err)
 		}
 
-		desc, err := imageStore.Resolve(ctx, targetImage.String())
+		imageStr := normalizedImage.String()
+
+		desc, err := imageStore.Resolve(ctx, imageStr)
 		if err != nil {
-			return fmt.Errorf("resolving image %q on local registry: %w", targetImage.String(), err)
+			return fmt.Errorf("resolving image %q on local registry: %w", imageStr, err)
 		}
 
-		digestString := desc.Digest.String()
-		found := false
-		for _, digest := range imgOpts.AllowedDigests {
-			if digestString == digest {
-				found = true
+		imageDigest := desc.Digest.String()
+
+		for _, allowedGadget := range imgOpts.AllowedGadgets {
+			if strings.Contains(allowedGadget, "sha") {
+				parts := strings.Split(allowedGadget, ":")
+				if len(parts) != 2 {
+					return fmt.Errorf("%s has badly formed sha", allowedGadget)
+				}
+
+				allowedDigest := parts[1]
+				found = allowedDigest == imageDigest
+			} else if strings.HasSuffix(allowedGadget, "*") {
+				found = strings.HasPrefix(imageStr, allowedGadget[:len(allowedGadget) - 1])
+			} else {
+				found = image == allowedGadget
+			}
 
+			if found {
 				break
 			}
```

Comment:
I'm wondering if it's faster just to compare against the digest in all cases without looking for the sha256 substring:

```suggestion
			// Check full match on digest or name
			if imageDigest == allowedGadget || imageStr == allowedGadget {
				found = true
				break
			}

			// Check prefix match
			if allowedGadget[len(allowedGadget)-1] == '*' {
				if strings.HasPrefix(imageStr, allowedGadget[:len(allowedGadget)-1]) {
					found = true
					break
				}
			}
```

Reply:
oh right. I missed that imageDigest should be calculated like `normalizedImage.Name() + "@" + desc.Digest.String()`. 

With that change it works as expected:

```bash 
$ sudo -E ./ig run --allowed-gadgets 'ghcr.io/inspektor-gadget/gadget/trace_exec@sha256:631653090611d56a2e65c8b9d42cfc9cae9823f12f21adca3751b442b495c089' ghcr.io/inspektor-gadget/gadget/trace_exec:v0.29.0
RUNTIME.CONTAI… MNTNS_ID TIMESTA… PID      PPID     UID      GID      LOGINUID SESSION… RETVAL   ARGS_C… UPPER_… ARGS_S… COMM    ARGS
^C

$ sudo -E ./ig run --allowed-gadgets 'sha256:631653090611d56a2e65c8b9d42cfc9cae9823f12f21adca3751b442b495c089' ghcr.io/inspektor-gadget/gadget/trace_exec:v0.29.0
Error: fetching gadget information: initializing and preparing operators: instantiating operator "oci": ensuring image: ghcr.io/inspektor-gadget/gadget/trace_exec:v0.29.0 is not part of allowed gadgets: sha256:631653090611d56a2e65c8b9d42cfc9cae9823f12f21adca3751b442b495c089
```